akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala

```diff
@@ -12,56 +12,96 @@ import scala.concurrent.duration.Duration
 import scala.concurrent.duration.FiniteDuration
 
 private[akka] object ReceiveTimeout {
-  final val emptyReceiveTimeoutData: (Duration, Cancellable) = (Duration.Undefined, ActorCell.emptyCancellable)
+  import ActorCell.emptyCancellable
+
+  private final val empty: ReceiveTimeoutValue =
+    ReceiveTimeoutValue(Duration.Undefined, emptyCancellable, Duration.Undefined)
+
+  final case class ReceiveTimeoutValue(current: Duration, task: Cancellable, previous: Duration) {
+
+    def :+(latest: Duration): ReceiveTimeoutValue = latest match {
+      case null => this // user input check
+      case _    =>
+        val prev = current
+        copy(previous = prev, current = latest)
+    }
+
+    /** Returns true if `current` is different from `previous` duration. */
+    def changed: Boolean = current ne previous
+
+    def clear(): ReceiveTimeoutValue = copy(task = emptyCancellable)
+
+    /* read-only functions */
+    def nonEmpty: Boolean = this ne empty
+    def taskIsEmpty: Boolean = task eq emptyCancellable
+    def taskNonEmpty: Boolean = !taskIsEmpty
```

Comment:
just isolating the internal logic to keep `ReceiveTimeout` cleaner. Nicer for testing in isolation, though no additional tests added yet.